lib/core/src/Cardano/Wallet/Primitive/Types/Tx.hs

```diff
@@ -392,11 +416,91 @@ instance Buildable a => Buildable (WithDirection a) where
+-- be submitted to the node.
+--
+-- Construct it with either 'sealedTxFromCardano' or 'sealedTxFromBytes'.
+data SealedTx = SealedTx
+    { cardanoTx :: InAnyCardanoEra Cardano.Tx
+    , serialisedTx :: ByteString
+    } deriving stock (Generic)
+
+instance Show SealedTx where
+    show (SealedTx (InAnyCardanoEra _era tx) _) = show tx
+
+instance Eq SealedTx where
+    SealedTx (InAnyCardanoEra e1 _) a == SealedTx (InAnyCardanoEra e2 _) b =
+        case testEquality e1 e2 of
+            Just Refl -> a == b
+            Nothing -> False
+
+instance NFData SealedTx where
+    rnf = rnf . show  -- fixme: temp fix
+
+-- | Construct a 'SealedTx' from a "Cardano.Api" transaction.
+sealedTxFromCardano :: InAnyCardanoEra Cardano.Tx -> SealedTx
+sealedTxFromCardano tx = SealedTx tx (cardanoTxToBytes tx)
+  where
+    cardanoTxToBytes :: InAnyCardanoEra Cardano.Tx -> ByteString
+    cardanoTxToBytes (InAnyCardanoEra _era tx') = Cardano.serialiseToCBOR tx'
+
+-- | Deserialise a Cardano transaction. The transaction can be in the format of
+-- any era. This function will try the most recent era first, then
+-- previous eras until 'ByronEra'.
+cardanoTxFromBytes
+    :: AnyCardanoEra -- ^ Most recent era
+    -> ByteString -- ^ Serialised transaction
+    -> Either DecoderError (InAnyCardanoEra Cardano.Tx)
+cardanoTxFromBytes maxEra bs = asum $ map snd $ filter (withinEra maxEra . fst)
+    [ (anyCardanoEra AlonzoEra, InAnyCardanoEra AlonzoEra <$> deserialiseFromCBOR (Cardano.AsTx Cardano.AsAlonzoEra) bs)
+    , (anyCardanoEra MaryEra, InAnyCardanoEra MaryEra <$> deserialiseFromCBOR (Cardano.AsTx Cardano.AsMaryEra) bs)
+    , (anyCardanoEra AllegraEra, InAnyCardanoEra AllegraEra <$> deserialiseFromCBOR (Cardano.AsTx Cardano.AsAllegraEra) bs)
+    , (anyCardanoEra ShelleyEra, InAnyCardanoEra ShelleyEra <$> deserialiseFromCBOR (Cardano.AsTx Cardano.AsShelleyEra) bs)
+    , (anyCardanoEra ByronEra, InAnyCardanoEra ByronEra <$> deserialiseFromCBOR (Cardano.AsTx Cardano.AsByronEra) bs)
+    ]
+  where
+    asum xs = case partitionEithers xs of
+        (_, (a:_)) -> Right a
+        ((e:_), []) -> Left e
+        ([], []) -> error "impossible"
+
+withinEra :: AnyCardanoEra -> AnyCardanoEra -> Bool
+withinEra (AnyCardanoEra AlonzoEra) _ = True
+withinEra (AnyCardanoEra MaryEra) (AnyCardanoEra AlonzoEra) = False
+withinEra (AnyCardanoEra MaryEra) (AnyCardanoEra _) = True
+withinEra (AnyCardanoEra AllegraEra) (AnyCardanoEra AlonzoEra) = False
+withinEra (AnyCardanoEra AllegraEra) (AnyCardanoEra MaryEra) = False
+withinEra (AnyCardanoEra AllegraEra) (AnyCardanoEra _) = True
+withinEra (AnyCardanoEra ShelleyEra) (AnyCardanoEra AlonzoEra) = False
+withinEra (AnyCardanoEra ShelleyEra) (AnyCardanoEra MaryEra) = False
+withinEra (AnyCardanoEra ShelleyEra) (AnyCardanoEra AllegraEra) = False
+withinEra (AnyCardanoEra ShelleyEra) (AnyCardanoEra _) = True
+withinEra (AnyCardanoEra ByronEra) (AnyCardanoEra ByronEra) = True
+withinEra (AnyCardanoEra ByronEra) _ = False
```

Comment:
Hi @paweljakubas. Just wondering, could this be expressed in terms of ordering for eras, with `<`?

Alternatively, this seems a bit shorter:

```hs
withinEra :: AnyCardanoEra -> AnyCardanoEra -> Bool
withinEra (AnyCardanoEra e1) (AnyCardanoEra e2) = case (e1, e2) of
    (AlonzoEra , _         ) -> True
    (MaryEra   , AlonzoEra ) -> False
    (MaryEra   , _         ) -> True
    (AllegraEra, AlonzoEra ) -> False
    (AllegraEra, MaryEra   ) -> False
    (AllegraEra, _         ) -> True
    (ShelleyEra, AlonzoEra ) -> False
    (ShelleyEra, MaryEra   ) -> False
    (ShelleyEra, AllegraEra) -> False
    (ShelleyEra, _         ) -> True
    (ByronEra  , ByronEra  ) -> True
    (ByronEra  , _         ) -> False
```

Reply:
There is no `Ord` on `AnyCardanoEra`, so that's what we're defining here.
The wildcard are because no era is higher that AlonzoEra, as an example.
Anyway I think it can be made shorter just by assigning a natural to each era.

Reply:
Assigning naturals would be morally equivalent to treating `AnyCardanoera` as an `Enum`, I suppose?